--- a/drivers/net/ethernet/atheros/ag71xx/ag71xx.h
+++ b/drivers/net/ethernet/atheros/ag71xx/ag71xx.h
@@ -55,6 +55,8 @@
 #define AG71XX_TX_RING_SPLIT		512
 #define AG71XX_TX_RING_DS_PER_PKT	DIV_ROUND_UP(AG71XX_TX_MTU_LEN, \
 						     AG71XX_TX_RING_SPLIT)
+#define AG71XX_HACK_WIFI_HEADROOM	128
+
 #define AG71XX_TX_RING_SIZE_DEFAULT	48
 #define AG71XX_RX_RING_SIZE_DEFAULT	128
 
--- a/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c
+++ b/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c
@@ -205,7 +205,7 @@ static void ag71xx_ring_rx_clean(struct
 
 static int ag71xx_buffer_offset(struct ag71xx *ag)
 {
-	int offset = NET_SKB_PAD;
+	int offset = AG71XX_HACK_WIFI_HEADROOM;
 
 	/*
 	 * On AR71xx/AR91xx packets must be 4-byte aligned.
